core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -394,12 +418,22 @@ public boolean process()
                 return false;
             }
 
-            // putIfAbsent will rehash automatically if rehash is needed, unless there isn't enough memory to do so.
-            // Therefore needRehash will not generally return true even if we have just crossed the capacity boundary.
-            while (lastPosition < positionCount && !needRehash()) {
-                // get the group for the current row
-                putIfAbsent(lastPosition, block);
-                lastPosition++;
+            int remainingPositions = positionCount - lastPosition;
+
+            long[] dummyGroupIds = new long[MAX_BATCH_SIZE];
+            while (remainingPositions != 0) {
+                int positionCountUntilRehash = maxFill - nextGroupId;
+                if (positionCountUntilRehash == 0) {
```

Comment:
Could you change it to:
```
if (positionCountUntilRehash < MAX_BATCH_SIZE) {
```
to avoid problem @lukasz-stec described below?

Then you wouldn't need `min` with `positionCountUntilRehash` below too

Reply:
That may increase the memory footprint, but as you wish